### PR TITLE
Correctly handle retries when both `preemptible` and `maxRetries` runtime parameters are set [BA-6102]

### PR DESCRIPTION
**Pull request by grsterin**

### CODE REVIEW

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala [hunk 1]
```diff
@@ -954,6 +954,43 @@ trait StandardAsyncExecutionActor
...
+      }
+
+    backendExecutionStatus flatMap {
+      case failed: FailedExecutionHandle =>
+        failed.kvPairsToSave match {
+          case Some(kvPairs) =>
+            val kvsFromBackend = kvPairs.map(kvPair => kvPair.key.key -> kvPair).toMap
+            val mergedKvs = (kvsFromPreviousAttempt ++ kvsFromBackend) map {
+              case (_, value) => value
+            }
```
Comment (mcovarr) [inline backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: I think this is equivalent
```
            val mergedKvs = (kvsFromPreviousAttempt.values ++ kvPairs)
```

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala [hunk 2]
```diff
@@ -681,13 +684,11 @@ class PipelinesApiAsyncBackendJobExecutionActor(override val standardParams: Sta
     }
   }
 
-  private def writeFuturePreemptedAndUnexpectedRetryCounts(p: Int, ur: Int): Future[Unit] = {
-    val updateRequests = Seq(
-      KvPut(KvPair(ScopedKey(workflowId, futureKvJobKey, PipelinesApiBackendLifecycleActorFactory.unexpectedRetryCountKey), ur.toString)),
-      KvPut(KvPair(ScopedKey(workflowId, futureKvJobKey, PipelinesApiBackendLifecycleActorFactory.preemptionCountKey), p.toString))
+  private def futurePreemptedAndUnexpectedRetryCountsToKvPairs(p: Int, ur: Int): Seq[KvPair] = {
```
Comment (mcovarr) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala]: TOL maybe 'next' rather than 'future'? 'future' sounds like `Future`.

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala (hunk 1 above)
Comment (grsterin) [reply backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: @mcovarr looks like this would simply append one sequence to another, but I wanted to enforce key uniqueness and precedence of values, which came from backend, over values from previous attempt, hence joining two `Map[String, KvPair]`

Comment (mcovarr) [reply backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: Ah right, as it says in the scaladoc. 🙂

File: backend/src/main/scala/cromwell/backend/async/ExecutionHandle.scala [hunk 3]
```diff
@@ -27,19 +28,32 @@ final case class PendingExecutionHandle[BackendJobId <: JobId, BackendRunInfo, B
...
 }
 
-final case class FailedNonRetryableExecutionHandle(throwable: Throwable, returnCode: Option[Int] = None) extends ExecutionHandle {
+sealed trait FailedExecutionHandle extends ExecutionHandle {
+  def kvPairsToSave: Option[Seq[KvPair]]
+}
+
+final case class FailedNonRetryableExecutionHandle(throwable: Throwable,
+                                                   returnCode: Option[Int] = None,
+                                                   override val kvPairsToSave: Option[Seq[KvPair]] = None) extends FailedExecutionHandle {
```
Comment (cjllanwarne) [inline backend/src/main/scala/cromwell/backend/async/ExecutionHandle.scala]: TOL: I'd probably not provide a default here. I kind of like the idea of forcing creators of these handles to explicitly opt-out of storing key/value pairs

File: centaur/src/main/resources/standardTestCases/papi_preemptible_and_max_retries.test [hunk 4]
```diff
@@ -0,0 +1,12 @@
+name: papi_preemptible_and_max_retries
+testFormat: workflowfailure
```
Comment (cjllanwarne) [inline centaur/src/main/resources/standardTestCases/papi_preemptible_and_max_retries.test]: TOL: is it possible to check that the appropriate call made three attempts before finally failing? I'm not 100% sure but there might be an existing centaur test option to check for that

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala [hunk 5]
```diff
@@ -698,10 +699,14 @@ class PipelinesApiAsyncBackendJobExecutionActor(override val standardParams: Sta
       case Valid(PreviousRetryReasons(p, ur)) =>
         val thisUnexpectedRetry = ur + 1
         if (thisUnexpectedRetry <= maxUnexpectedRetries) {
+          val preemptionAndUnexpectedRetryCountsKvPairs = nextAttemptPreemptedAndUnexpectedRetryCountsToKvPairs(p, thisUnexpectedRetry)
           // Increment unexpected retry count and preemption count stays the same
-          writeFuturePreemptedAndUnexpectedRetryCounts(p, thisUnexpectedRetry).map { _ =>
-            FailedRetryableExecutionHandle(StandardException(
-              errorCode, msg, jobTag, jobReturnCode, standardPaths.error), jobReturnCode)
+          Future.successful {
```
Comment (cjllanwarne) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala]: It looks like all of these cases now return `Future.successful`... can we instead update the method signature to return `ExecutionHandle` and remove the wrapping?

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala [hunk 6]
```diff
@@ -725,18 +730,24 @@ class PipelinesApiAsyncBackendJobExecutionActor(override val standardParams: Sta
         val taskName = s"${workflowDescriptor.id}:${call.localName}"
         val baseMsg = s"Task $taskName was preempted for the ${thisPreemption.toOrdinal} time."
 
-        writeFuturePreemptedAndUnexpectedRetryCounts(thisPreemption, ur).map { _ =>
+        Future.successful {
```
Comment (cjllanwarne) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala]: Same comment here - we could update the signature to return `ExecutionHandle`

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala [hunk 7]
```diff
@@ -954,6 +954,39 @@ trait StandardAsyncExecutionActor
...
   }
 
+  /**
+   * Merge key-value pairs from previous job execution attempt with incoming pairs from backend and store them in
+   * the database. In case when there are key-value pairs with the same key name among those from previous attempt and
+   * coming from backend, backend values have higher precedence.
+   * `FailedRetryCountKey` is handled separately outside of this method 
+   *
+   */
+  private def saveKvPairsForNextAttempt(backendExecutionStatus: Future[ExecutionHandle]): Future[ExecutionHandle] = {
```
Comment (cjllanwarne) [inline backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: I think I would:
* Change this method signature to `saveKvPairsForNextAttempt(backendExecutionStatus: ExecutionHandle): Future[ExecutionHandle]`
* Refactor the `flatMap` above into a for-comprehension over a chain of `Future[X]` methods.

UPDATE: I think actually this comment might be redundant given my next one...

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala [hunk 8]
```diff
@@ -954,6 +954,39 @@ trait StandardAsyncExecutionActor
...
+    val kvsFromPreviousAttempt = jobDescriptor.prefetchedKvStoreEntries
+      .collect {
+        // We exclude FailedRetryCountKey here since it will be incremented and saved separately, after we ensure
```
Comment (cjllanwarne) [inline backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: I originally envisioned that we would pass all of these values up to the engine - and allow _the engine_ to decide whether to write all of them, or none of them, depending on whether it restarts the job or not. Does that seem sensible to you?

Regardless, I think that if `FailedRetryCountKey` is now an engine concept (and I think it is...), that we should move the value out of the `backend` package and into the `engine` package - so that it's (hopefully) impossible for backends to try to write this in the first place.

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowExecutionActor.scala [hunk 9]
```diff
@@ -445,16 +463,93 @@ case class WorkflowExecutionActor(params: WorkflowExecutionActorParams)
...
+   * among those from previous attempt and coming from backend, backend values have higher precedence.
+   *
+   */
+  private def saveKvPairsForNextAttempt(jobKey: BackendJobDescriptorKey,
+                                        newKvJobKey: KvJobKey,
+                                        kvPairsFromPreviousAttemptOpt: Option[Seq[KvPair]],
+                                        kvPairsForNextAttemptOpt: Option[Seq[KvPair]],
+                                        kvPairFailedAttemptsOpt: Option[KvPair]): Unit = {
+
+    val optSeqKvPairToMap: Option[Seq[KvPair]] => Map[String, KvPair] = optSeq =>
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowExecutionActor.scala]: OOC why is this a `val` rather than a more straightforward `def`?

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala [hunk 10]
```diff
@@ -664,57 +672,72 @@ class PipelinesApiAsyncBackendJobExecutionActor(override val standardParams: Sta
...
       }
     }
 
     runStatus match {
-      case preemptedStatus: RunStatus.Preempted if preemptible => handlePreemption(preemptedStatus, returnCode)
+      case preemptedStatus: RunStatus.Preempted if preemptible => Future.successful(handlePreemption(preemptedStatus, returnCode))
       case _: RunStatus.Cancelled => Future.successful(AbortedExecutionHandle)
-      case failedStatus: RunStatus.UnsuccessfulRunStatus => handleFailedRunStatus(failedStatus, returnCode)
+      case failedStatus: RunStatus.UnsuccessfulRunStatus => Future.successful(handleFailedRunStatus(failedStatus, returnCode))
       case unknown => throw new RuntimeException(s"handleExecutionFailure not called with RunStatus.Failed or RunStatus.Preempted. Instead got $unknown")
```
Comment (mcovarr) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala]: `Future.failed`?

Comment (mcovarr) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala]: hmm actually why is everything getting wrapped in `Future`

**Comment by grsterin** [comment]
@cjllanwarne, @mcovarr, please pause further reviews for now. I don't like current implementation and I think that previous implementation actually may have been better. This needs additional discussion.

**Comment by grsterin** [comment]
> @cjllanwarne, @mcovarr, please pause further reviews for now. I don't like current implementation and I think that previous implementation actually may have been better. This needs additional discussion.

Moved handling of `preemptible` and `maxRetries` attributes from Engine to backend's `StandardAsyncExecutionActor`.

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala [hunk 11]
```diff
@@ -649,72 +652,92 @@ class PipelinesApiAsyncBackendJobExecutionActor(override val standardParams: Sta
...
+            kvPairsToSave = Option(preemptionAndUnexpectedRetryCountsKvPairs)
+          )
         }
         else {
-          Future.successful(FailedNonRetryableExecutionHandle(StandardException(
-            errorCode, errorMessage, jobTag, jobReturnCode, standardPaths.error), jobReturnCode))
+          FailedNonRetryableExecutionHandle(
+            StandardException(errorCode, errorMessage, jobTag, jobReturnCode, standardPaths.error),
+            jobReturnCode,
+            None
```
Comment (cjllanwarne) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala]: I think we still want to preserve the `kvPairs`? In case the memory doubling logic decides to retry the job?

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala [hunk 12]
```diff
@@ -649,72 +652,92 @@ class PipelinesApiAsyncBackendJobExecutionActor(override val standardParams: Sta
...
+          )
         }
       case Invalid(_) =>
-        Future.successful(FailedNonRetryableExecutionHandle(StandardException(
-          errorCode, errorMessage, jobTag, jobReturnCode, standardPaths.error), jobReturnCode))
+        FailedNonRetryableExecutionHandle(
+          StandardException(errorCode, errorMessage, jobTag, jobReturnCode, standardPaths.error),
+          jobReturnCode,
+          None
```
Comment (cjllanwarne) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala]: I think we still want to preserve the `kvPairs`? In case the memory doubling logic decides to retry the job?

**Comment by cjllanwarne** [comment]
This feels like the kind of thing that would really benefit from a design diagram/mini document, if you think you might be able to put one together? The overlapping ways to retry and starting to get really hard to follow (a) how they all happen, (b) when they all happen, (c) where they all happen and (d) how we make sure the KV store is correctly carried foward across them all...

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala (hunk 11 above)
Comment (grsterin) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala]: In this case we don't need to increment neither `preempted` counter nor `unexpectedRetry` counter, so we just preserve values from previous attempt (this logic is implemented in `StandardAsyncExecutionActor.retryElseFail`).
We need to pass new key-value pairs in ExecutionHandle only if we want to introduce some new parameters, or change value of the existing ones.

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala (hunk 12 above)
Comment (grsterin) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala]: In this case we don't need to increment neither `preempted` counter nor `unexpectedRetry` counter, so we just preserve values from previous attempt (this logic is implemented in `StandardAsyncExecutionActor.retryElseFail`).
We only need to pass new key-value pairs in ExecutionHandle if we want to introduce some new parameters, or change value of the existing ones.

**Comment by grsterin** [comment]
> This feels like the kind of thing that would really benefit from a design diagram/mini document, if you think you might be able to put one together? The overlapping ways to retry and starting to get really hard to follow (a) how they all happen, (b) when they all happen, (c) where they all happen and (d) how we make sure the KV store is correctly carried foward across them all...

BA-6139

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala [hunk 13]
```diff
@@ -917,43 +917,106 @@ trait StandardAsyncExecutionActor
...
+          }.toMap
+          case None => Map.empty[String, KvPair]
+        }
+
+        val maxRetriesNotReachedYet = previousFailedRetries < maxRetries
+        failedRetryableOrNonRetryable match {
+          case failed: FailedNonRetryableExecutionHandle if maxRetriesNotReachedYet =>
+            val currentMemoryMultiplier = jobDescriptor.key.memoryMultiplier
+            (retryWithMoreMemory, memoryRetryFactor) match {
+              case (true, Some(multiplier)) => refineV[GreaterEqualOne](currentMemoryMultiplier.value * multiplier.value) match {
```
[the remark this answers is not in the document]
Comment (grsterin) [reply backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: Agreed. Btw, multiplier value is already being verified while when reading from config.

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala [hunk 14]
```diff
@@ -917,43 +917,106 @@ trait StandardAsyncExecutionActor
...
         }
       case _ => backendExecutionStatus
     }
   }
 
+  private def saveAttrsAndRetry(failedExecHandle: FailedExecutionHandle,
+                        kvPrev: Map[String, KvPair],
+                        kvNext: Map[String, KvPair],
+                        incFailedCount: Boolean,
+                        memoryMultiplier: Option[GreaterEqualRefined] = None): Future[FailedRetryableExecutionHandle] = {
```
[the remark this answers is not in the document]
Comment (grsterin) [reply backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: No, since `type GreaterEqualRefined = Refined[Double, GreaterEqualOne]` and `refineMv[GreaterEqualOne](some_double_value)` returns value of type `GreaterEqualRefined`

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala [hunk 15]
```diff
@@ -917,43 +917,106 @@ trait StandardAsyncExecutionActor
...
+                  )
+                )
+                case Right(newMultiplier) => saveAttrsAndRetry(failed, kvsFromPreviousAttempt, kvsForNextAttempt, incFailedCount = true, Option(newMultiplier))
+              }
+              case (_, _) => saveAttrsAndRetry(failed, kvsFromPreviousAttempt, kvsForNextAttempt, incFailedCount = true)
             }
-            case (_, _) => FailedRetryableExecutionHandle(failed.throwable, failed.returnCode, currentMemoryMultiplier)
-          }
+          case failedNonRetryable: FailedNonRetryableExecutionHandle => Future.successful(failedNonRetryable)
+          case failedRetryable: FailedRetryableExecutionHandle => saveAttrsAndRetry(failedRetryable, kvsForNextAttempt, kvsForNextAttempt, incFailedCount = false)
```
[the remark this answers is not in the document]
Comment (grsterin) [reply backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: This is because if we get `FailedRetryableExecutionHandle` here, that means that job is retryable for some reason which is known and checked by backend, but which doesn't affect failed retries counter yet.
For example, we end up here, if preemption occurred, but the max number of preemption retries has not been exceeded yet (this is checked in the particular backend actor implementation, `PipelinesApiAsyncBackendJobExecutionActor`). In this case there will be already incremented preemption counter inside the `kvsForNextAttempt` parameter. 
But if number of preemptions has already reached the configured maximum, then we'll have `FailedNonRetryableExecutionHandle` here and we'll start checking the failed retries counter.

[the remark this answers is not in the document]
Comment (grsterin) [reply backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: A doc describing retry logic will be created under a separate ticket: BA-6139
